### ISSUE TITLE
Dynamic mapping does not work with more than one type

### ISSUE THREAD

**Issue by AColmant** [elastic/elasticsearch#16348]
Hello guys,

This config worked in ES v1.7.2 but since i tried to use the configuration in ES v2.1 I have a error... So :

I use this mapping :
```json
PUT /my_index/_mapping/_default_
{
    "_default_": {
        "dynamic_date_formats" : [
            "yyyy-MM-dd HH:mm:ss",
            "yyyy-MM-dd HH:mm:ss||yyyyMMdd-HH:mm:ss||yyyyMMdd-HH:mm:ss.SSS"
        ],
        "dynamic_templates": [
            {
                "template_1" : {
                    "match" : "*",
                    "match_mapping_type" : "string",
                    "mapping" : {
                        "type" : "string",
                        "index" : "not_analyzed"
                    }
                }
            }, {
                "template_2" : {
                    "match" : "*@expire",
                    "match_mapping_type" : "date",
                    "mapping" : {
                        "type" : "string",
                        "index" : "not_analyzed"
                    }
                }
            }
        ]
    }
}
```

And after a first type insert, I get the expected result. But if I try to insert a second or third type like :
```json
POST /my_index/my_type
{
   "source":"my_source",
   "country":"my_country",
   "company":"my_company",
   "server":"my_server",
   "@timestamp":"2016-01-01 00:07:10"
}
```

Result return is :
```json
{
   "error": {
      "root_cause": [
         {
            "type": "mapper_parsing_exception",
            "reason": "failed to parse [@timestamp]"
         }
      ],
      "type": "mapper_parsing_exception",
      "reason": "failed to parse [@timestamp]",
      "caused_by": {
         "type": "illegal_argument_exception",
         "reason": "Invalid format: \"2016-01-01 00:07:10\" is malformed at \" 00:07:10\""
      }
   },
   "status": 400
}
```

If I try to insert it on the first type, this is working... Perhaps bug, or just my mistake but i do not understand.

FYI: if i swap the order of the type inserted, it is always the first that it work correctly.

**Comment by jimferenczi** [comment]
@AColmant thanks for the report, you are right this is a bug and you can find the fix for ES v2.3.0 via https://github.com/elastic/elasticsearch/issues/15568 and https://github.com/elastic/elasticsearch/issues/15138.

**Comment by AColmant** [comment]
Alright ! Sorry for the issue duplicate and thanks for your job.
